crypto/thread/arch/thread_posix.c

```diff
@@ -0,0 +1,290 @@
+    ret = thread->routine(thread->data);
+    CRYPTO_THREAD_SET_STATE(thread, CRYPTO_THREAD_FINISHED);
+    thread->retval = ret;
+    ossl_crypto_condvar_broadcast(thread->condvar);
+    ossl_crypto_mutex_unlock(thread->statelock);
+
+    return NULL;
+}
+
+int ossl_crypto_thread_native_spawn(CRYPTO_THREAD *thread)
+{
+    int ret;
+    pthread_attr_t attr;
+    pthread_t *handle;
+
+    handle = OPENSSL_zalloc(sizeof(*handle));
+    if (handle == NULL)
+        goto fail;
+
+    pthread_attr_init(&attr);
+    if (!thread->joinable)
+        pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);
+    ret = pthread_create(handle, &attr, thread_start_thunk, thread);
+    pthread_attr_destroy(&attr);
+
+    if (ret != 0)
+        goto fail;
+
+    thread->handle = handle;
+    return 1;
+
+fail:
+    thread->handle = NULL;
+    OPENSSL_free(handle);
+    return 0;
+}
+
+int ossl_crypto_thread_native_join(CRYPTO_THREAD *thread, CRYPTO_THREAD_RETVAL *retval)
+{
+    void *thread_retval;
+    pthread_t *handle;
+    uint64_t req_state_mask;
+
+    if (thread == NULL)
+        return 0;
+
+    req_state_mask = CRYPTO_THREAD_TERMINATED | CRYPTO_THREAD_JOINED;
+
+    ossl_crypto_mutex_lock(thread->statelock);
+    if (CRYPTO_THREAD_GET_STATE(thread, req_state_mask))
+        goto pass;
+    while (!CRYPTO_THREAD_GET_STATE(thread, CRYPTO_THREAD_FINISHED))
+        ossl_crypto_condvar_wait(thread->condvar, thread->statelock);
+
+    handle = (pthread_t *) thread->handle;
+    if (handle == NULL)
+        goto fail;
+
+    if (pthread_join(*handle, &thread_retval) != 0)
+        goto fail;
+
+    /*
+     * Join return value may be non-NULL when the thread has been cancelled,
+     * as indicated by thread_retval set to PTHREAD_CANCELLED.
+     */
+    if (thread_retval != NULL)
+        goto fail;
+
+pass:
+    if (retval != NULL)
+        *retval = thread->retval;
+
+    CRYPTO_THREAD_UNSET_ERROR(thread, CRYPTO_THREAD_JOINED);
+    CRYPTO_THREAD_SET_STATE(thread, CRYPTO_THREAD_JOINED);
+    ossl_crypto_mutex_unlock(thread->statelock);
+    return 1;
+
+fail:
+    CRYPTO_THREAD_SET_ERROR(thread, CRYPTO_THREAD_JOINED);
+    ossl_crypto_mutex_unlock(thread->statelock);
+    return 0;
+}
+
+int ossl_crypto_thread_native_terminate(CRYPTO_THREAD *thread)
+{
+    uint64_t mask;
+    pthread_t *handle;
+
+    mask = CRYPTO_THREAD_FINISHED;
+    mask |= CRYPTO_THREAD_TERMINATED;
+    mask |= CRYPTO_THREAD_JOINED;
+
+    if (thread == NULL)
+        return 1;
+
+    if (thread->handle == NULL || CRYPTO_THREAD_GET_STATE(thread, mask))
+        goto terminated;
+
+    handle = thread->handle;
+    if (pthread_cancel(*handle) != 0) {
+        CRYPTO_THREAD_SET_ERROR(thread, CRYPTO_THREAD_TERMINATED);
+        return 0;
+    }
+
+    thread->handle = NULL;
+    OPENSSL_free(handle);
+
+ terminated:
+    CRYPTO_THREAD_UNSET_ERROR(thread, CRYPTO_THREAD_TERMINATED);
+    CRYPTO_THREAD_SET_STATE(thread, CRYPTO_THREAD_TERMINATED);
```

Comment:
Similarly here - don't we need to be holding `statelock` to call `CRYPTO_THREAD_SET_STATE`?